examples/bzlmod/hello_world/MODULE.bazel

```diff
@@ -0,0 +1,18 @@
+module(
+    name = "hello_world",
+    toolchains_to_register = ["@rust_linux_x86_64__x86_64-unknown-linux-gnu//:toolchain"],
+    version = "1.0",
+)
+
+bazel_dep(name = "rules_rust", version = "0.9.0")
+
+local_path_override(
+    module_name = "rules_rust",
+    path = "../../..",
+)
+
+rust_toolchains = use_extension("@rules_rust//:extensions.bzl", "toolchains")
```

Comment:
I think the toolchain registration can be done in rules_rust's MODULE.bazel file, so that the example module can just depend on rules_rust and get the toolchain for free?

Reply:
Which toolchains are available depends on what parameters the user passes to `rust_toolchains.toolchain`. For example, passing `extra_target_triples` generates additional toolchains.

At least, this would be the case if I were to add all the parameters to the tag class that can be forwarded to `rust_register_toolchains`.

Reply:
This does seem to be a significant downside to using bzlmod here. Where previously the user of rules_rust didn't have to know which toolchains were registered, they now have to at the least have the following boilerplate in their MODULE.bazel, as well as know how the options passed to `rust_toolchains.toolchain` map to generated repositories:

```
use_repo(
    rust_toolchains,
    "rust_linux_x86_64__x86_64-unknown-linux-gnu",
    "rust_darwin_aarch64__aarch64-apple-darwin",
    "rust_linux_aarch64__aarch64-unknown-linux-gnu",
    "rust_darwin_x86_64__x86_64-apple-darwin",
    "rust_windows_x86_64__x86_64-pc-windows-msvc",
    "rust_freebsd_x86_64__x86_64-unknown-freebsd",
)

register_toolchains(
    "@rust_linux_x86_64__x86_64-unknown-linux-gnu//:toolchain",
    "@rust_darwin_aarch64__aarch64-apple-darwin//:toolchain",
    "@rust_linux_aarch64__aarch64-unknown-linux-gnu//:toolchain",
    "@rust_darwin_x86_64__x86_64-apple-darwin//:toolchain",
    "@rust_windows_x86_64__x86_64-pc-windows-msvc//:toolchain",
    "@rust_freebsd_x86_64__x86_64-unknown-freebsd//:toolchain",
)
```

I can't think of a way to avoid this currently.

Reply:
It seems we lack of a way to dynamically register generated toolchains. @Wyverald Is it possible to support `register_toolchain` in module extension implementation?

Reply:
it's possible but we decided not to. I think it's a bit late to introduce it at this point for 6.0. Worst comes to worst, we could have a module called "rules_rust_toolchains" that contains nothing but the above in its MODULE.bazel file.

Reply:
This may be a pretty significant performance regression - each of these toolchains is in a tar which is hundreds of megabytes, so forcing the user to download 6 of these just to get a toolchain definitions, 5 of which they won't use, is pretty bad.

We could re-structure so that the toolchain definitions live in separate repositories from the actual toolchains, to avoid the proactive downloads, but I wanted to highlight that this is a significant API difference from Bazel.

Reply:
That's very true. The setup you mentioned is the same as the Java toolchains, so at least there's precedent.

I can think of another potential solution to this. We can make the `toolchains` extension generate a "hub" repo (let's say called "toolchains_hub"), which has nothing but a list of aliases pointing to toolchains that the user requested. And then, we could make rules_rust always register everything in `@toolchains_hub//:all`. That would mean whatever the user requests via tags gets registered, without the user having to do anything. Something like the following:

```bzl
# rules_rust/MODULE.bazel
toolchains = use_extension("//:extensions.bzl", "toolchains")
use_repo(toolchains, "toolchains_hub")
register_toolchains("@toolchains_hub//:all")

# rules_rust/extensions.bzl
toolchains = module_extension(_toolchains_impl, tag_classes = ...)
def _toolchains_impl(ctx):
  toolchain_repos = _somehow_get_repos_from_tags(ctx)
  for toolchain_repo in toolchain_repos: rust_toolchain_repo(**toolchain_repo)
  toolchain_hub_repo(name="toolchains_hub", repos=[tr["name"] for tr in toolchain_repos])
rust_toolchain_repo = repository_rule(...)
toolchain_hub_repo = repository_rule(_toolchain_hub_repo_impl, attrs={"repos": attr.string_list()})
def _toolchain_hub_repo_impl(ctx):
  ctx.file("WORKSPACE")
  ctx.file("BUILD", "\n".join(["alias(name='%s', actual='@%s//:toolchain')" % (repo,repo) for repo in ctx.attr.repos]))

# my_project/MODULE.bazel
bazel_dep(name="rules_rust", version="...")
rust_toolchains = use_extension("//:extensions.bzl", "toolchains")
rust_toolchains.toolchain(edition="2021")
```

Reply:
This would work well, as long as using `:all` works correctly for toolchain registration. I'll try it out later.

Reply:
Regarding the performance regression, is this any different from the current situation? A lot of unused toolchains get registered currently.

Reply:
Even without Bzlmod, all toolchains get registered in WORKSPACE will be fetched for toolchain resolution, so I think this is the same as the current situation.

Reply:
> Regarding the performance regression, is this any different from the current situation? A lot of unused toolchains for different platforms get registered currently.

Aah good point, we already split out the meta-toolchain from the expensive repo: https://github.com/bazelbuild/rules_rust/blob/cb9ca1b817febcaf8edc28d160866be759825c31/rust/repositories.bzl#L495-L510

Sorry for the noise!

Reply:
It looks like toolchain registration doesn't work with `:all`, since I get the following error when I try to use it:

```
ERROR: /home/cameron/Repos/rules_rust/examples/bzlmod/hello_world/BUILD.bazel:5:12: While resolving toolchains for target //:hello_world: No matching toolchains found for types @rules_rust~override//rust:toolchain_type.
To debug, rerun with --toolchain_resolution_debug='@rules_rust~override//rust:toolchain_type'
```

I've updated the PR to use a hub repo anyway, since it does cut down on the boilerplate.

Reply:
If registering with `:all` isn't working, it's a bug. We have tests (such as https://cs.opensource.google/bazel/bazel/+/master:src/test/java/com/google/devtools/build/lib/skyframe/ToolchainsForTargetsTest.java;l=219;drc=16c4076cd39f43f3ccabbdeb6ab407c2063b038c) ensuring this behavior.

Do you have more information to help us debug? Check the BUILD file you generated to see whether it actually contains toolchains, etc.?

Reply:
The BUILD file contains a bunch of aliases to toolchains:

```
alias(name='rust_darwin_aarch64__aarch64-apple-darwin', actual='@rust_darwin_aarch64__aarch64-apple-darwin//:toolchain', visibility = ["//visibility:public"])
alias(name='rust_linux_aarch64__aarch64-unknown-linux-gnu', actual='@rust_linux_aarch64__aarch64-unknown-linux-gnu//:toolchain', visibility = ["//visibility:public"])
alias(name='rust_darwin_x86_64__x86_64-apple-darwin', actual='@rust_darwin_x86_64__x86_64-apple-darwin//:toolchain', visibility = ["//visibility:public"])
alias(name='rust_windows_x86_64__x86_64-pc-windows-msvc', actual='@rust_windows_x86_64__x86_64-pc-windows-msvc//:toolchain', visibility = ["//visibility:public"])
alias(name='rust_freebsd_x86_64__x86_64-unknown-freebsd', actual='@rust_freebsd_x86_64__x86_64-unknown-freebsd//:toolchain', visibility = ["//visibility:public"])
alias(name='rust_linux_x86_64__x86_64-unknown-linux-gnu', actual='@rust_linux_x86_64__x86_64-unknown-linux-gnu//:toolchain', visibility = ["//visibility:public"])
```

The issue seems to be with the interaction between `:all` and aliases. I've filed a bug report at https://github.com/bazelbuild/bazel/issues/16298.

Reply:
I've updated this to put all the toolchain proxies into one repository. This does change the public interface, currently does break non-bzlmod usage and does not include rust-analyzer support, but the idea is there.

Reply:
This hub repository now duplicates the toolchain proxies instead of replaces them, so the public API does not change.